routes/check_api_version_test.go

```diff
@@ -6,61 +6,76 @@ import (
 	"net/http/httptest"
 	"testing"
 
-	"github.com/gocardless/draupnir/version"
 	"github.com/stretchr/testify/assert"
 )
 
-func TestCheckApiVersionWithNoVersionHeader(t *testing.T) {
-	recorder := httptest.NewRecorder()
-	req := httptest.NewRequest("GET", "/foo", nil)
-
-	CheckAPIVersion(
-		func(w http.ResponseWriter, h *http.Request) {
-			t.Fatal("this route should never be called")
+func TestCheckApiVersion(t *testing.T) {
+	testCases := []struct {
+		name          string
+		headerVersion string
+		apiError      APIError
+		code          int
+	}{
+		{
+			"when version matches, calls handler",
+			"1.1.0",
+			APIError{},
+			http.StatusAccepted,
+		},
+		{
+			"when minor is lower, calls handler",
+			"1.0.0",
+			APIError{},
+			http.StatusAccepted,
+		},
+		{
+			"when minor is higher, responds with error",
+			"1.2.0",
+			invalidApiVersion("1.2.0"),
+			http.StatusBadRequest,
+		},
+		{
+			"when header major version is different, responds with error",
+			"0.1.0",
+			invalidApiVersion("0.1.0"),
+			http.StatusBadRequest,
+		},
+		{
+			"when header is missing, responds with error",
+			"",
+			missingApiVersion,
+			http.StatusBadRequest,
 		},
-	)(recorder, req)
-
-	assert.Equal(t, recorder.Code, http.StatusBadRequest)
-
-	var response APIError
-	err := json.NewDecoder(recorder.Body).Decode(&response)
-	if err != nil {
-		t.Fatal(err)
 	}
-	assert.Equal(t, response, missingApiVersion)
-}
 
-func TestCheckApiVersionWithMismatchingVersionHeader(t *testing.T) {
-	recorder := httptest.NewRecorder()
-	req := httptest.NewRequest("GET", "/foo", nil)
-	req.Header["Draupnir-Version"] = []string{"0.0.0"}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			recorder := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/foo", nil)
 
-	CheckAPIVersion(
-		func(w http.ResponseWriter, h *http.Request) {
-			t.Fatal("this route should never be called")
-		},
-	)(recorder, req)
+			if tc.headerVersion != "" {
```

Comment:
[hmac] Do we need this conditional?

Reply:
Yep, so that we can test when the header isn't there. Simply providing an empty string means you get an invalid error, rather than a 'missing header' response.

Reply:
Makes sense 👍